packages/flutter_migrate/lib/src/compute.dart

```diff
@@ -0,0 +1,1014 @@
+  for (final String path in _skippedFiles) {
+    if (path.replaceAll('/', fileSystem.path.separator) == localPath) {
+      return true;
+    }
+  }
+  if (_skippedFiles.contains(localPath)) {
+    return true;
+  }
+  for (final String dir in _skippedDirectories) {
+    if (localPath.startsWith(
+        '${dir.replaceAll('/', fileSystem.path.separator)}${fileSystem.path.separator}')) {
+      return true;
+    }
+  }
+  if (blacklistPrefixes != null) {
+    for (final String prefix in blacklistPrefixes) {
+      if (localPath.startsWith('$prefix${fileSystem.path.separator}')) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+const List<String> _skippedMergeFileExt = <String>[
+  // Don't merge image files
+  '.png',
+  '.jpg',
+  '.jpeg',
+  '.gif',
+  '.bmp',
+  '.svg',
+  // Don't merge compiled artifacts and executables
+  '.jar',
+  '.so',
+  '.exe',
+];
+
+const Set<String> _alwaysMigrateFiles = <String>{
+  '.metadata', // .metadata tracks key migration information.
+  'android/gradle/wrapper/gradle-wrapper.jar',
+  // Always add .gitignore back in even if user-deleted as it makes it
+  // difficult to migrate in the future and the migrate tool enforces git
+  // usage.
+  '.gitignore',
+};
+
+/// True for files that should not be merged. Typically, images and binary files.
+bool _skippedMerge(String localPath) {
+  for (final String ext in _skippedMergeFileExt) {
+    if (localPath.endsWith(ext) && !_alwaysMigrateFiles.contains(localPath)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Data class holds the common context that is used throughout the steps of a migrate computation.
+class MigrateContext {
+  MigrateContext({
+    required this.migrateResult,
+    required this.flutterProject,
+    required this.blacklistPrefixes,
+    required this.logger,
+    required this.verbose,
+    required this.fileSystem,
+    required this.status,
+    required this.migrateUtils,
+    this.baseProject,
+    this.targetProject,
+  });
+
+  MigrateResult migrateResult;
+  FlutterProject flutterProject;
+  Set<String> blacklistPrefixes;
+  Logger logger;
+  bool verbose;
+  FileSystem fileSystem;
+  Status status;
+  MigrateUtils migrateUtils;
+
+  MigrateBaseFlutterProject? baseProject;
+  MigrateTargetFlutterProject? targetProject;
+}
+
+String getLocalPath(String path, String basePath, FileSystem fileSystem) {
+  return path.replaceFirst(basePath + fileSystem.path.separator, '');
+}
+
+String platformToSubdirectoryPrefix(SupportedPlatform platform) {
+  switch (platform) {
+    case SupportedPlatform.android:
+      return 'android';
+    case SupportedPlatform.ios:
+      return 'ios';
+    case SupportedPlatform.linux:
+      return 'linux';
+    case SupportedPlatform.macos:
+      return 'macos';
+    case SupportedPlatform.web:
+      return 'web';
+    case SupportedPlatform.windows:
+      return 'windows';
+    case SupportedPlatform.fuchsia:
+      return 'fuchsia';
+    case SupportedPlatform.root:
```

Comment:
According to the docs this is a "Special platform to represent the root project directory" (which already seems confusing to me, as it's conflating directories and platforms). In what context would returning a non-existent directory here rather than erroring out make sense?

Reply:
I'll refactor this to not mix these concepts!